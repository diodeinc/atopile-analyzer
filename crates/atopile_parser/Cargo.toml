[package]
name = "atopile_parser"
version = "0.1.0"
edition = "2021"

[dependencies]
chumsky = "0.9.3"
insta = "1.41.1"
serde = { version = "1.0.214", features = ["derive", "rc"] }

[dev-dependencies]
insta = "1.34.0"
glob = "0.3"
