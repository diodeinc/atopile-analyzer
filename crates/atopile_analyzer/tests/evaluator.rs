use atopile_analyzer::evaluator::Evaluator;
use atopile_parser::AtopileSource;
use std::fs;
use std::path::PathBuf;

macro_rules! create_evaluator_test {
    ($name:ident) => {
        #[test]
        fn $name() {
            let input = fs::read_to_string(concat!(
                "tests/resources/corpus/",
                stringify!($name),
                ".ato"
            ))
            .unwrap();
            let normalized_input = input.replace("\r\n", "\n");

            let file_path = concat!("tests/resources/corpus/", stringify!($name), ".ato");
            let path_buf = PathBuf::from(file_path);
            
            let (source, _errors) = AtopileSource::new(
                normalized_input.to_string(),
                path_buf.clone(),
            );
            
            let mut evaluator = Evaluator::new();
            let result = evaluator.evaluate(&source);

            insta::assert_debug_snapshot!(result);
        }
    };
}

create_evaluator_test!(vdivs);
create_evaluator_test!(resistors);
create_evaluator_test!(transistors);
create_evaluator_test!(bma400);
