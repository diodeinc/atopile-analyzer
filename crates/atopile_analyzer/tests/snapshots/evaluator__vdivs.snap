---
source: crates/atopile_analyzer/tests/evaluator.rs
expression: result
---
EvaluatorState {
    instances: {
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_sda",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            kind: Component,
            attributes: {
                Symbol(
                    "mpn",
                ): String(
                    "generic_resistor",
                ),
                Symbol(
                    "type",
                ): String(
                    "resistor",
                ),
                Symbol(
                    "value",
                ): Physical(
                    "10kohm Â± 20%",
                ),
                Symbol(
                    "package",
                ): String(
                    "0402",
                ),
                Symbol(
                    "designator_prefix",
                ): String(
                    "R",
                ),
            },
            children: {
                Symbol(
                    "p1",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_sda",
                        ),
                        Symbol(
                            "p1",
                        ),
                    ],
                },
                Symbol(
                    "p2",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_sda",
                        ),
                        Symbol(
                            "p2",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_bottom",
                ),
                Symbol(
                    "p2",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_bottom",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            kind: Component,
            attributes: {
                Symbol(
                    "mpn",
                ): String(
                    "generic_resistor",
                ),
                Symbol(
                    "type",
                ): String(
                    "resistor",
                ),
                Symbol(
                    "package",
                ): String(
                    "0402",
                ),
                Symbol(
                    "designator_prefix",
                ): String(
                    "R",
                ),
            },
            children: {
                Symbol(
                    "p1",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_bottom",
                        ),
                        Symbol(
                            "p1",
                        ),
                    ],
                },
                Symbol(
                    "p2",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_bottom",
                        ),
                        Symbol(
                            "p2",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "out",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            kind: Module,
            attributes: {},
            children: {
                Symbol(
                    "out",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "out",
                        ),
                    ],
                },
                Symbol(
                    "top",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "top",
                        ),
                    ],
                },
                Symbol(
                    "r_top",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_top",
                        ),
                    ],
                },
                Symbol(
                    "bottom",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "bottom",
                        ),
                    ],
                },
                Symbol(
                    "r_bottom",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_bottom",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "bottom",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_scl",
                ),
                Symbol(
                    "p1",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            kind: Module,
            attributes: {},
            children: {
                Symbol(
                    "r_scl",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_scl",
                        ),
                    ],
                },
                Symbol(
                    "r_sda",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_sda",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            instance_path: [
                Symbol(
                    "p1",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "top",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_top",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            kind: Component,
            attributes: {
                Symbol(
                    "mpn",
                ): String(
                    "generic_resistor",
                ),
                Symbol(
                    "type",
                ): String(
                    "resistor",
                ),
                Symbol(
                    "package",
                ): String(
                    "0402",
                ),
                Symbol(
                    "designator_prefix",
                ): String(
                    "R",
                ),
            },
            children: {
                Symbol(
                    "p1",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_top",
                        ),
                        Symbol(
                            "p1",
                        ),
                    ],
                },
                Symbol(
                    "p2",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "tests/resources/corpus/vdivs.ato",
                        module_name: Symbol(
                            "_VDiv",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_top",
                        ),
                        Symbol(
                            "p2",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            instance_path: [],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            kind: Component,
            attributes: {
                Symbol(
                    "mpn",
                ): String(
                    "generic_resistor",
                ),
                Symbol(
                    "type",
                ): String(
                    "resistor",
                ),
                Symbol(
                    "designator_prefix",
                ): String(
                    "R",
                ),
            },
            children: {
                Symbol(
                    "p1",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "Resistor",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "p1",
                        ),
                    ],
                },
                Symbol(
                    "p2",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "Resistor",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "p2",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_top",
                ),
                Symbol(
                    "p2",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_sda",
                ),
                Symbol(
                    "p1",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_scl",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            kind: Component,
            attributes: {
                Symbol(
                    "mpn",
                ): String(
                    "generic_resistor",
                ),
                Symbol(
                    "type",
                ): String(
                    "resistor",
                ),
                Symbol(
                    "value",
                ): Physical(
                    "10kohm Â± 20%",
                ),
                Symbol(
                    "package",
                ): String(
                    "0402",
                ),
                Symbol(
                    "designator_prefix",
                ): String(
                    "R",
                ),
            },
            children: {
                Symbol(
                    "p1",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_scl",
                        ),
                        Symbol(
                            "p1",
                        ),
                    ],
                },
                Symbol(
                    "p2",
                ): InstanceRef {
                    module: ModuleRef {
                        source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                        module_name: Symbol(
                            "I2CPullup",
                        ),
                    },
                    instance_path: [
                        Symbol(
                            "r_scl",
                        ),
                        Symbol(
                            "p2",
                        ),
                    ],
                },
            },
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "Resistor",
                ),
            },
            instance_path: [
                Symbol(
                    "p2",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_sda",
                ),
                Symbol(
                    "p2",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_top",
                ),
                Symbol(
                    "p1",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "/home/ubuntu/repos/atopile-analyzer/crates/atopile_analyzer/tests/resources/corpus/resistors.ato",
                module_name: Symbol(
                    "I2CPullup",
                ),
            },
            instance_path: [
                Symbol(
                    "r_scl",
                ),
                Symbol(
                    "p2",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
        InstanceRef {
            module: ModuleRef {
                source_path: "tests/resources/corpus/vdivs.ato",
                module_name: Symbol(
                    "_VDiv",
                ),
            },
            instance_path: [
                Symbol(
                    "r_bottom",
                ),
                Symbol(
                    "p1",
                ),
            ],
        }: Instance {
            type_ref: ModuleRef {
                source_path: "",
                module_name: Symbol(
                    "",
                ),
            },
            kind: Port,
            attributes: {},
            children: {},
            connections: [],
            reference_designator: None,
        },
    },
}
